lib/modules/manager/pip-compile/extract.ts

```diff
@@ -216,19 +220,27 @@ export async function extractAllPackageFiles(
   // This needs to go in reverse order to handle transitive dependencies
   for (const packageFile of [...result].reverse()) {
     for (const reqFile of packageFile.managerData?.requirementsFiles ?? []) {
-      let sourceFile: PackageFile | undefined = undefined;
+      let sourceFiles: PackageFile[] | undefined = undefined;
       if (fileMatches.includes(reqFile)) {
-        sourceFile = lockFileSources.get(reqFile);
+        sourceFiles = lockFileSources.get(reqFile);
       } else if (packageFiles.has(reqFile)) {
-        sourceFile = packageFiles.get(reqFile);
+        sourceFiles = [packageFiles.get(reqFile)!];
       }
-      if (!sourceFile) {
+      if (!sourceFiles) {
         logger.warn(
           `pip-compile: ${packageFile.packageFile} references ${reqFile} which does not appear to be a requirements file managed by pip-compile`,
         );
         continue;
       }
-      sourceFile.lockFiles!.push(...packageFile.lockFiles!);
+      for (const sourceFile of sourceFiles) {
+        let merged = [...packageFile.lockFiles!];
+        for (const lockFile of [...sourceFile.lockFiles!].reverse()) {
+          if (!merged.includes(lockFile)) {
+            merged = [lockFile, ...merged];
+          }
```

Comment:
```suggestion
         merged.add(lockfile);
```

recreating the array is too expensive, if the set doesn't work use `merged.unshift(lockFile)` which will move it to the start of the array and use the set for contains check only.